apps/glusterfs/examiner.go

```diff
@@ -0,0 +1,153 @@
+//
+// Copyright (c) 2018 The heketi Authors
+//
+// This file is licensed to you under your choice of the GNU Lesser
+// General Public License, version 3 or any later version (LGPLv3 or
+// later), or the GNU General Public License, version 2 (GPLv2), in all
+// cases as published by the Free Software Foundation.
+//
+
+package glusterfs
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/boltdb/bolt"
+	"github.com/heketi/heketi/executors"
+)
+
+type ExaminerMode string
+
+const (
+	OnDemandExaminer ExaminerMode = "ondemand"
+	OfflineExaminer  ExaminerMode = "offline"
+)
+
+type Examiner struct {
+	db        *bolt.DB
+	executor  executors.Executor
+	optracker *OpTracker
+	mode      ExaminerMode
+}
+
+type NodeData struct {
+	NodeHeketiID string
+	VolumeInfo   *executors.VolInfo
+}
+
+type ClusterData struct {
+	ClusterHeketiID string
+	NodesData       []NodeData
+}
+
+type GlusterStateExaminationResponse struct {
+	HeketiDB Db            `json:"heketidb"`
+	Report   []string      `json:"report"`
+	Clusters []ClusterData `json:"clusters"`
+}
+
+func (examiner Examiner) fetchClusterData(cluster ClusterEntry, heketidb Db) (clusterdata ClusterData, errorstrings []string) {
+
+	clusterdata.ClusterHeketiID = cluster.Info.Id
+
+	for _, node := range cluster.Info.Nodes {
+		nodeEntry := heketidb.Nodes[node]
+		var nodedata NodeData
+		nodedata.NodeHeketiID = nodeEntry.Info.Id
+
+		// Volume Info
+		volinfo, err := examiner.executor.VolumesInfo(nodeEntry.ManageHostName())
+		if err != nil {
+			errorstrings = append(errorstrings, fmt.Sprintf("could not fetch data from node %v", nodeEntry.ManageHostName()))
+		} else {
+			nodedata.VolumeInfo = volinfo
+		}
+
+		// Append whatever information we could gather, minimum is the node ID
+		clusterdata.NodesData = append(clusterdata.NodesData, nodedata)
+	}
+
+	return
+
+}
+
+func matchVolumes(heketidb Db, cdata ClusterData) (errorstrings []string) {
+
+	var heketiVolList []string
+
+	for _, volume := range heketidb.Clusters[cdata.ClusterHeketiID].Info.Volumes {
+		heketiVolList = append(heketiVolList, heketidb.Volumes[volume].Info.Name)
+	}
+
+	for _, node := range cdata.NodesData {
+		var volList []string
+		if node.VolumeInfo == nil {
+			errorstrings = append(errorstrings, fmt.Sprintf("heketi volume list could not be compared with volume list of node %v due to missing info", node.NodeHeketiID))
+			continue
+		}
+		for _, volume := range node.VolumeInfo.Volumes.VolumeList {
+			volList = append(volList, volume.VolumeName)
+		}
+		if !reflect.DeepEqual(heketiVolList, volList) {
```

Comment:
If you are using lists here, do they need to be sorted at some point, or not, or did I miss it?